File: go_solution/011_020/threeSumClosest/threeSumClosest.go
package threeSumClosest

import "sort"

// 16. 3Sum Closest
// 思路：和 3Sum 相似的思路
func threeSumClosest(nums []int, target int) int {
	if len(nums) == 3 {
		return nums[0] + nums[1] + nums[2]
	}
	var (
		rlt   []int
		delta int64 = 1<<63 - 1
		perv        = 0
	)
	sort.Ints(nums)
	for idx, v := range nums {
		if len(rlt) != 0 && v == perv {
			continue
		}
		l, r := idx+1, len(nums)-1
		for l < r {
			if calDelta(nums[l]+nums[r]+v, target) < delta {
				rlt = []int{nums[l] + nums[r] + v}
				delta = calDelta(nums[l]+nums[r]+v, target)
				perv = v
			} else if nums[l]+nums[r]+v > target {
				r--
			} else {
				l++
			}
		}
	}
	return rlt[0]
}

func calDelta(val1 int, val2 int) int64 {
	return abs(int64(val1 - val2))
}

func abs(n int64) int64 {
	y := n >> 63
	return (n ^ y) - y
}

File: java_solution/problem011_020/ThreeSumClosest.java
package problem011_020;

import java.util.*;

/**
 * 016. Three Sum Closest
 * 思路同015. Three Sum
 */
public class ThreeSumClosest {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        // 指定最大的三个数之和为初始值
        int result = nums[nums.length - 3] + nums[nums.length - 2] + nums[nums.length - 1];
        for (int i = 0; i < nums.length - 1; i++){
            if(i == 0 || nums[i] > nums[i-1]) {
                int l = i + 1, r = nums.length - 1;
                while(l < r) {
                    int sum = nums[i] + nums[l] + nums[r];
                    if (Math.abs(sum - target) < Math.abs(result - target)) {
                        result = sum;
                    }
                    if (sum == target) {
                        return sum;
                    } else if (sum > target) {
                        r--;
                    } else {
                        l++;
                    }
                }
            }
        }
        return result;
    }
}
